[5.5.x] Add check for rollback of latest unstarted phase

bernardjkim opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR updates the check for `gravity plan rollback` command. The check now ensures that phases of an operation are being rolled back in order. This check can be bypassed by providing the `--force` flag.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Ports https://github.com/gravitational/gravity/pull/2184

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [x] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
**Verify rollback phase fails if out of order and `--force` flag is not provided**

**Verify rollback phase executes if out of order and `--force` flag is provided**

**Verify latest phase can still be rolled back**

**Verify plan can be rolled back completely**

bernardjkim committed:
Add check for rollback of latest unstarted phase

bernardjkim commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,42 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+	if !dryRun && !latestRollback(plan.Phases, phaseID) {
```
Rolling back an operation as a dry run does not change the state of operation phases. So dry run rollbacks would always fail this check. For now, simple workaround was to just ignore this check during dry runs.

a-palchikov commented on lib/fsm/fsm.go:
```diff
@@ -257,7 +257,7 @@ func (f *FSM) RollbackPhase(ctx context.Context, p Params) error {
 		return trace.Wrap(err)
 	}
-	err = CanRollback(plan, p.PhaseID)
+	err = CanRollback(plan, p.PhaseID, p.DryRun)
```
No need to pass `dryRun` into CanRollback since it only makes sense to evaluate in case of an actual rollback:
```go
if !p.DryRun {
	err = CanRollback(plan, p.PhaseID)
	if err != nil {
		if !p.Force {
			return trace.Wrap(err)
		}
		f.WithError(err).Warn("Forcing rollback.")
	}
}
```

a-palchikov commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,42 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 }
 
+// latestRollback returns true if the phase specified by phaseID is next in line
+// to be rolledback.
```
nit: rolled back

a-palchikov commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,42 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+		switch {
+		// if we run into an unstarted phase, all later phases should also be unstarted
+		case phase.IsUnstarted():
+			break
```
this is equivalent to the empty `case` clause (as it terminates the innermost `switch`):
```go
switch {
case phase.IsUnstarted():
case phase.IsInProgress(), phase.IsFailed(), phase.IsCompleted():
...
}
```
if you needed to terminate the loop - use a label.

bernardjkim replied:
Oops. Yes, we would want to terminate the loop here.

a-palchikov commented on lib/fsm/utils_test.go:
```diff
@@ -148,3 +148,123 @@ func (s *FSMUtilsSuite) TestDiffPlanNoPrevious(c *check.C) {
+	}
+	for _, tc := range tests {
+		comment := check.Commentf(tc.comment)
+		c.Assert(latestRollback(tc.phases, tc.phaseID), check.Equals, tc.expected, comment)
```
I would actually write a test for the exported `CanRollback` rather than exercising the private latestCallback.
Also, I'm not seeing the case where there's a rolled back phase is at the end (i.e. prior to an unstarted one) - how is it supposed to behave in this case?

bernardjkim committed:
Address PR feedback

- Ignore CanRollback check if dry run.
- Use label to break loop.
- Add extra test case.

bernardjkim committed:
Remove expected empty string

r0mant commented on lib/fsm/fsm.go:
```diff
@@ -257,12 +257,14 @@ func (f *FSM) RollbackPhase(ctx context.Context, p Params) error {
-	if err != nil {
-		if !p.Force {
-			return trace.Wrap(err)
+	if !p.DryRun {
```
Add a comment on why these checks aren't executed in dry-run mode?

r0mant commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,44 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+			return true
+		}
+		switch {
+		// if we run into an unstarted phase, all later phases should also be unstarted
```
That's technically not true, a phase can be executed "out of band" using plan execute command so I don't think we can make this assumption.

bernardjkim replied:
Ah, thanks for pointing that out. Will update.

r0mant commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,44 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+L:
+	for _, phase := range phases {
+		if latestRollback(phase.Phases, phaseID) {
```
Why do we need recursion here? I would make the algorithm very simple, all we need to do is to "unwrap" the plan (we have the method called "FlattenPlan" somewhere, there's also a method called GetLeafPhases on the plan itself IIRC) and then check if any of the phases that appear after phaseID have been rolled back or are unstarted. Sounds simple enough and easy to reason about, what do you think?

r0mant replied:
And I think this is good. If I'm not mistaken, rolling back top-level phases is not implemented properly at the moment, as in it starts rollback top-down, not in reverse order. Or it may be disabled completely, I don't recall.

bernardjkim replied:
Oh I see. Okay, we can update to only allow rollback of leaf phases.

bernardjkim committed:
Address feedback

- Top level phases require --force to be rolled back
- Take into account that phases can be execed out of band

knisbet commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,54 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+	leafPhases := plan.GetLeafPhases()
+
+	// TODO: Rollback of non-leaf phases may currently not be implemented
```
Nit: "Rollback of non-leaf phases may currently not be implemented" -> Rollback of non-leaf phases is not supported. 

Although I think this get's a bit strange in gravity, because there is a rollback command that can rollback entirely. So may be difficult to convey that this applies specifically here.

knisbet commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,54 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+	// properly. Roll back starts top-down, and not in reverse order.
+	if !containsPhase(leafPhases, phaseID) {
+		return trace.BadParameter(
+			"attempting to rollback non-leaf phase %q", phase.ID)
```
Nit: I don't think a customer would understand why they receive an error like this. So I would consider going something more along the lines of:

`return trace.BadParameter("unable to rollback non-leaf phase, only leaf phases can be rolled back.").AddField("phase", phase.ID)`

knisbet commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,54 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+// latestRollback returns true if the phase specified by phaseID is next in line
+// to be rolled back.
+func latestRollback(phases []storage.OperationPhase, phaseID string) bool {
```
I'm a bit worried about this algorithm to determine dependence, especially since we kind of support parallel phase execution which I'm probably going to turn on for upgrades of large clusters. By looking at the latest phase from the flattened structure, it creates a dependency that doesn't necessarily exist between phases. IE if in an upgrade I want to rollback / retry bootstrap on node-1, I have to rollback nodes 2-800. 

I think identifying when a rollback can happen would be a closer fit if any depends on operations are in progress, failed, completed. So it's the same readiness to move forwards as it is to move backwords. Although this is a much much more complicated algorithm with the way dependencies are structures, because they can point to a group phase or leaf phase.

r0mant replied:
> IE if in an upgrade I want to rollback / retry bootstrap on node-1, I have to rollback nodes 2-800.

This is exactly the situation that we want to prevent customers from getting in, unless they know what they're doing. In a recent support scenario, someone rolled back one of masters without rolling back any of subsequent phases which put cluster in a bad state. For "power users" there should be a --force flag to override this check and allow you to rollback anything. @bernardjkim - btw I don't see this flag implemented?

r0mant replied:
Nevermind, I see the --force flag is handled above.

bernardjkim replied:
@r0mant @knisbet 
I've updated the implementation in 0bf4934 to be more accommodating towards rolling back parallel phases.

bernardjkim committed:
Address feedback

- Update docs and error msg
- Use phase.HasSubphases to check if leaf phase

bernardjkim committed:
Update rollback check implementation

- Use `requires` field to find dependent phases and check if all
depedents have been rolled back if they have been started.

bernardjkim committed:
Add test for parallel phase rollback

bernardjkim commented on lib/fsm/utils.go:
```diff
@@ -41,9 +41,54 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+	// TODO: Rollback of non-leaf phases may currently not be implemented
+	// properly. Roll back starts top-down, and not in reverse order.
+	if !containsPhase(leafPhases, phaseID) {
```
Looks like we can make it even simpler and just check if `phase.HasSubPhases`

r0mant commented on lib/fsm/utils_test.go:
```diff
@@ -148,3 +149,248 @@ func (s *FSMUtilsSuite) TestDiffPlanNoPrevious(c *check.C) {
+							{
+								ID:       "/nodes/node-3",
+								State:    storage.OperationPhaseStateCompleted,
+								Requires: []string{"/nodes-node-2"},
```
Typo? `/nodes-node-2` -> `/nodes/node-2`

bernardjkim replied:
Oops, good catch.

r0mant commented on lib/fsm/utils.go:
```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+// getRequired constructs the initial set of required phases. This set includes
+// the phase specified by phaseID and its parent phases.
+func getRequired(phases []storage.OperationPhase, phaseID string) map[string]struct{} {
```
I would update the godoc comment with an example of what it returns, to simplify understanding. I.e.

```
Given plan like:

/init
/masters
  * /node-1
    * /system-upgrade
  * /node-2

for "/masters/node-1/system-upgrade" phase id it will return:

["/masters", "/masters/node-1", "/masters/node-1/system-upgrade"]
```

Also, why does this return a map, doesn't the order matter?

bernardjkim replied:
Order doesn't matter, we just need the set to check if any other phases 'requires' any of the phases in this set.

r0mant commented on lib/fsm/utils.go:
```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+}
+
+// dependentsRolledBackHelper is a recursive helper function that returns true
+// if all phases that are dependent on the required set is unstarted or has
```
*are* unstarted, *have* been

r0mant commented on lib/fsm/utils.go:
```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+			if !phase.IsUnstarted() && !phase.IsRolledBack() {
+				return false
+			}
+			required[phase.ID] = struct{}{}
```
Hmm, do we need to add this phase to the "required" set as well? I wonder if it'll be simpler to just leave the original set as it is, this way the algo will be just one dependency "deep", right? So for example, in case of the following plan:

```
/init
/masters <--depends on /init
/gc <-- depends on /masters
```

when you try to rollback /init and /masters and /gc are in progress, it would only tell you to rollback /masters first. Then you try to rollback /masters and it will tell you to rollback /gc first.

We could probably even get rid of recursion in this case, just iterate over all phases and see if any of them depend on those in the "required" set.

What do you think?

bernardjkim replied:
That approach would definitely be a lot simpler, but I'm not sure if the check will function properly if a phase is executed out of band. Since phases can be executed out of order, wouldn't something like this be a valid state of the plan?
```
/Init { state: completed }
/masters { state: unstarted, requires: { /init } }
/gc { state: completed, requires: { /masters } }
```
In this situation all phases that require /init are either un-started or rolled back.
Would /gc in this case even have been able to be executed?

r0mant replied:
Yes, you're right. Let's keep it as-is then, but add a comment here (as it's also a side-effect to the passed in parameter which is not great) and a test case for this scenario.

r0mant commented on lib/fsm/utils_test.go:
```diff
@@ -148,3 +149,248 @@ func (s *FSMUtilsSuite) TestDiffPlanNoPrevious(c *check.C) {
+	}{
+		{
+			comment: "Rollback latest phase",
+
```
Nit: Extra new line.

r0mant commented on lib/fsm/utils_test.go:
```diff
@@ -148,3 +149,248 @@ func (s *FSMUtilsSuite) TestDiffPlanNoPrevious(c *check.C) {
+		},
+		{
+			comment: "Top level phase has dependent phases that have not been rolled back",
+			plan: &storage.OperationPlan{
```
Can you add a few more test cases with more complicated plans, i.e. multiple-level deep phases, multiple "requires", etc., more like our real plan. E.g.

```
/init
/masters <-- depends on /init
  * /node-1
    * /drain
    * /system-upgrade <-- depends on /masters/node-1/drain
  * /node-2
    * /drain
    * /system-upgrade <-- depends on /masters/node-2/drain
/nodes <-- depends on /masters
  * /node-3
    * /drai
    * /system-upgrade
/runtime <-- depends on /masters
  * /monitoring
  * /site
/gc <-- depends on /runtime
```

... you get the idea... and try different scenarios.

r0mant reviewed (commented):
I actually quite like the algorithm, just have one simplification suggestion. @a-palchikov @knisbet can you guys take a look too.

bernardjkim committed:
Typos and error msg changes

bernardjkim commented on lib/fsm/utils.go:
```diff
@@ -43,9 +43,85 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
+
+	if !dependentsRolledBack(plan, phase.ID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
```
The check short circuits on the first dependent phase that needs to be rolled back, so we would only be able to display that first phase. I think this could be more confusing than helpful if we only display that one phase.

We could extend the check for the purpose of collecting all dependent phases. I don't think it would affect performance that much, and I think better ux might be worth it.

r0mant replied:
Yeah let's do it :+1:

bernardjkim replied:
Updated in 447efa2

Example output:
```
[vagrant@node-1 installer]$ sudo ./gravity plan rollback --phase /checks
[ERROR]: phase /checks cannot be rolled back because some phases that depend on it haven't been rolled back yet. Please rollback the following phases first:

        /pre-update
        /bootstrap

you can pass --force flag to override this check and force phase /checks rollback
```

a-palchikov reviewed (approved):
As long as other comments have been address - especially the extended error message listing the phases that prevent the rollback of a particular phase.

bernardjkim committed:
Collect and display phases that require rollback

bernardjkim committed:
Add additional test cases

bernardjkim committed:
Update phase state evaluation

- If all subphases are either unstarted or rolled back, the phase is
considered rolled back.
- Add withState. Only leaf phases should specify state.
- Capitalize error message.

bernardjkim committed:
Add test cases for GetState

bernardjkim committed:
Update local phase state on remote rollback

bernardjkim commented on lib/fsm/fsm.go:
```diff
@@ -329,13 +329,28 @@ You can redeploy upgrade agents on all cluster nodes using "./gravity agent depl
+	if err == nil {
+		// if the remote phase rollback is successful, we need to mark it in our local database
+		// because etcd might not be available to synchronize the changes back to us
+		err = f.ChangePhaseState(ctx,
```
`gravity rollback` would fail in the middle of rolling back a plan whenever a phase was rolled back remotely. 
```
[vagrant@node-1 installer]$ sudo ./gravity rollback
You are about to rollback the following operation:
Type                 ID                                       State                  Created
----                 --                                       -----                  -------
operation_update     c8e539eb-23ef-4e7b-999c-e022ce04963b     update_in_progress     2020-10-07 01:07
Consider checking the operation plan and using --dry-run flag first to see which actions will be performed.
You can suppress this warning in future by providing --confirm flag.
Proceed? (yes/no):
yes
Wed Oct  7 01:14:20 UTC Rolling back "/masters/node-3/untaint" locally
Wed Oct  7 01:14:24 UTC Rolling back "/masters/node-3/endpoints" locally
Wed Oct  7 01:14:31 UTC Rolling back "/masters/node-3/uncordon" locally
Wed Oct  7 01:14:39 UTC Rolling back "/masters/node-3/taint" locally
Wed Oct  7 01:14:45 UTC Rolling back "/masters/node-3/system-upgrade" on node node-3
        Still rolling back "/masters/node-3/system-upgrade" on node node-3 (10 seconds elapsed)
        Still rolling back "/masters/node-3/system-upgrade" on node node-3 (20 seconds elapsed)
        Still rolling back "/masters/node-3/system-upgrade" on node node-3 (30 seconds elapsed)
        Still rolling back "/masters/node-3/system-upgrade" on node node-3 (40 seconds elapsed)
Wed Oct  7 01:15:31 UTC operation(update(c8e539eb-23ef-4e7b-999c-e022ce04963b), cluster=dev.test, created=2020-10-07 01:07) finished in 1 minute
[ERROR]: phase /masters/node-3/drain cannot be rolled back because some phases that depend on it haven't been rolled back yet. Please rollback the following phases first:
        /masters/node-3/system-upgrade
you can pass --force flag to override this check and force phase /masters/node-3/drain rollback
```
This was because `rollbackPhaseRemotely` was not updating the local plan after rolling back a phase.
